File: ecps_uv/trust/secure_transport.py
"""
Secure transport wrapper for ECPS UV SDK.

This module provides a secure transport wrapper that adds security
features to any transport implementation, using identity forwarding
instead of per-request signing for better performance.
"""

import asyncio
import base64
import json
import logging
import time
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Callable, Dict, List, Optional, Tuple, Union

from ecps_uv.trust.trust import Principal, TrustLevel, TrustProvider
from ecps_uv.trust.identity_forwarding import (
    IdentityContext,
    IdentityForwardingManager,
    ForwardedRequest
)

logger = logging.getLogger("ecps_uv.trust.secure_transport")


@dataclass
class SecureMessage:
    """Wrapper for secure messages."""
    # Original message in serialized form
    message: bytes
    # Type name of the original message
    message_type: str
    # Message signature (if signed)
    signature: Optional[bytes] = None
    # Encrypted symmetric key (if encrypted)
    encrypted_key: Optional[bytes] = None
    # Initialization vector for symmetric encryption
    iv: Optional[bytes] = None
    # Security token (e.g., JWT)
    security_token: Optional[str] = None
    # ID of the sender
    sender_id: Optional[str] = None
    # Message timestamp
    timestamp: int = field(default_factory=lambda: int(time.time_ns()))

    def to_dict(self) -> Dict[str, Any]:
        """
        Convert to dictionary.
        
        Returns:
            Dictionary representation
        """
        result = {
            "message": base64.b64encode(self.message).decode("ascii"),
            "message_type": self.message_type,
            "timestamp": self.timestamp,
        }
        
        if self.signature:
            result["signature"] = base64.b64encode(self.signature).decode("ascii")
            
        if self.encrypted_key:
            result["encrypted_key"] = base64.b64encode(self.encrypted_key).decode("ascii")
            
        if self.iv:
            result["iv"] = base64.b64encode(self.iv).decode("ascii")
            
        if self.security_token:
            result["security_token"] = self.security_token
            
        if self.sender_id:
            result["sender_id"] = self.sender_id
            
        return result

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "SecureMessage":
        """
        Create from dictionary.
        
        Args:
            data: Dictionary representation
            
        Returns:
            SecureMessage instance
        """
        # Convert base64 fields back to bytes
        message = base64.b64decode(data["message"])
        
        signature = None
        if "signature" in data:
            signature = base64.b64decode(data["signature"])
            
        encrypted_key = None
        if "encrypted_key" in data:
            encrypted_key = base64.b64decode(data["encrypted_key"])
            
        iv = None
        if "iv" in data:
            iv = base64.b64decode(data["iv"])
            
        return cls(
            message=message,
            message_type=data["message_type"],
            signature=signature,
            encrypted_key=encrypted_key,
            iv=iv,
            security_token=data.get("security_token"),
            sender_id=data.get("sender_id"),
            timestamp=data.get("timestamp", int(time.time_ns())),
        )


class SecureTransport:
    """Secure transport wrapper using identity forwarding."""

    def __init__(
        self,
        transport: Any,
        trust_provider: TrustProvider,
        identity_forwarding_manager: Optional[IdentityForwardingManager] = None,
        identity_context: Optional[IdentityContext] = None,
    ):
        """
        Initialize the secure transport.
        
        Args:
            transport: The transport to wrap
            trust_provider: The trust provider
            identity_forwarding_manager: Manager for identity forwarding
            identity_context: Current identity context (if established)
        """
        self.transport = transport
        self.trust_provider = trust_provider
        self.identity_forwarding_manager = identity_forwarding_manager
        self.identity_context = identity_context
        self.serializer = getattr(transport, "serializer", None)

    def set_identity_context(self, identity_context: IdentityContext) -> None:
        """
        Set the identity context for this transport.
        
        Args:
            identity_context: The identity context to use
        """
        self.identity_context = identity_context

    async def publish(
        self,
        topic: str,
        message: Any,
        qos: Optional[Dict[str, Any]] = None
    ) -> None:
        """
        Publish a message with identity forwarding.
        
        Args:
            topic: Topic to publish to
            message: Message to publish
            qos: Quality of service parameters
        """
        # Skip security if trust level is none
        if self.trust_provider.trust_level == TrustLevel.NONE:
            return await self.transport.publish(topic, message, qos)
        
        # Check if we have identity context
        if not self.identity_context:
            raise ValueError("No identity context available for secure transport")
        
        # Check if identity context is still valid
        if self.identity_context.is_expired:
            raise ValueError("Identity context has expired")
        
        # Create forwarded request instead of signing each message
        if self.identity_forwarding_manager:
            forwarded_request = self.identity_forwarding_manager.create_forwarded_request(
                message, self.identity_context
            )
            
            # Convert to dictionary for transport
            forwarded_dict = forwarded_request.to_dict()
            
            # Publish forwarded request
            await self.transport.publish(topic, forwarded_dict, qos)
        else:
            # Fallback to legacy secure message approach
            if not self.serializer:
                raise ValueError("Transport does not provide a serializer")
            
            # Serialize the original message
            message_bytes = await self.serializer.serialize(message, use_json=False)
            
            # Create secure message with identity context
            secure_msg = SecureMessage(
                message=message_bytes,
                message_type=type(message).__name__,
                sender_id=self.identity_context.identity.id,
                security_token=self.identity_context.session_id,  # Use session ID as token
            )
            
            # Convert to dictionary for transport
            secure_dict = secure_msg.to_dict()
            
            # Publish secure message
            await self.transport.publish(topic, secure_dict, qos)

    async def subscribe(
        self,
        topic: str,
        handler: Callable,
        message_type: Any = None,
        qos: Optional[Dict[str, Any]] = None
    ) -> None:
        """
        Subscribe to a topic with identity forwarding validation.
        
        Args:
            topic: Topic to subscribe to
            handler: Callback function for received messages
            message_type: Expected message type
            qos: Quality of service parameters
        """
        # Skip security if trust level is none
        if self.trust_provider.trust_level == TrustLevel.NONE:
            return await self.transport.subscribe(topic, handler, message_type, qos)
        
        # Create forwarded request handler
        async def forwarded_handler(forwarded_dict: Dict[str, Any]) -> None:
            try:
                # Check if this is a forwarded request
                if "identity_context" in forwarded_dict and self.identity_forwarding_manager:
                    # Handle forwarded request
                    forwarded_request = ForwardedRequest(
                        payload=forwarded_dict["payload"],
                        identity_context=IdentityContext(
                            identity=forwarded_dict["identity_context"]["identity"],
                            principal_id=forwarded_dict["identity_context"]["principal_id"],
                            session_id=forwarded_dict["identity_context"]["session_id"],
                            established_at=datetime.fromisoformat(forwarded_dict["identity_context"]["established_at"]),
                            expires_at=datetime.fromisoformat(forwarded_dict["identity_context"]["expires_at"]),
                            capabilities=set(forwarded_dict["identity_context"]["capabilities"]),
                            attributes=forwarded_dict["identity_context"]["attributes"],
                        ),
                        request_id=forwarded_dict["request_id"],
                        timestamp=datetime.fromisoformat(forwarded_dict["timestamp"]),
                        service_chain=forwarded_dict["service_chain"],
                    )
                    
                    # Validate forwarded request
                    is_valid, error_reason = await self.identity_forwarding_manager.validate_forwarded_request(
                        forwarded_request, None, f"subscribe:{topic}"
                    )
                    
                    if not is_valid:
                        logger.error(f"Forwarded request validation failed: {error_reason}")
                        return
                    
                    # Authorize if required
                    if self.trust_provider.trust_level >= TrustLevel.AUTHORIZED:
                        authorized, reason = await self.identity_forwarding_manager.authorize_forwarded_request(
                            forwarded_request, "subscribe", topic
                        )
                        if not authorized:
                            logger.error(f"Forwarded request authorization failed: {reason}")
                            return
                    
                    # Call original handler with the payload
                    await handler(forwarded_request.payload)
                    
                else:
                    # Handle legacy secure message
                    if not self.serializer:
                        raise ValueError("Transport does not provide a serializer")
                    
                    # Convert dictionary to SecureMessage
                    secure_msg = SecureMessage.from_dict(forwarded_dict)
                    
                    # Validate session ID if available
                    if (self.trust_provider.trust_level >= TrustLevel.AUTHENTICATED and
                        secure_msg.security_token and self.identity_forwarding_manager):
                        context = self.identity_forwarding_manager.get_context_by_session(secure_msg.security_token)
                        if not context:
                            logger.error("Invalid or expired session ID")
                            return
                    
                    # Deserialize original message
                    orig_message = await self.serializer.deserialize(
                        secure_msg.message,
                        message_type or secure_msg.message_type
                    )
                    
                    # Call original handler with the deserialized message
                    await handler(orig_message)
                
            except Exception as e:
                logger.error(f"Error handling forwarded message: {e}")

        # Subscribe with forwarded handler
        await self.transport.subscribe(topic, forwarded_handler, Dict[str, Any], qos)

    async def request(
        self, 
        service: str, 
        request: Any, 
        timeout: float = 5.0, 
        qos: Optional[Dict[str, Any]] = None
    ) -> Any:
        """
        Send a request with security features.
        
        Args:
            service: Service name
            request: Request message
            timeout: Request timeout
            qos: Quality of service parameters
            
        Returns:
            Response message
        """
        # Not implementing security for request/response yet
        # In a real implementation, we would wrap the request in a SecureMessage
        # and unwrap the response
        return await self.transport.request(service, request, timeout, qos)

    async def stream_request(
        self, 
        service: str, 
        request: Any, 
        handler: Callable, 
        timeout: float = 5.0, 
        qos: Optional[Dict[str, Any]] = None
    ) -> None:
        """
        Send a streaming request with security features.
        
        Args:
            service: Service name
            request: Request message
            handler: Handler for streaming responses
            timeout: Request timeout
            qos: Quality of service parameters
        """
        # Not implementing security for streaming request yet
        # In a real implementation, we would wrap the request in a SecureMessage
        # and unwrap the responses
        await self.transport.stream_request(service, request, handler, timeout, qos)

    async def register_service(
        self, 
        service: str, 
        handler: Callable, 
        request_type: Any, 
        response_type: Any, 
        qos: Optional[Dict[str, Any]] = None
    ) -> None:
        """
        Register a service with security features.
        
        Args:
            service: Service name
            handler: Service handler
            request_type: Request message type
            response_type: Response message type
            qos: Quality of service parameters
        """
        # Not implementing security for service registration yet
        # In a real implementation, we would unwrap incoming SecureMessage objects
        # and wrap outgoing responses
        await self.transport.register_service(
            service, 
            handler, 
            request_type, 
            response_type, 
            qos
        )

    async def register_stream_service(
        self, 
        service: str, 
        handler: Callable, 
        request_type: Any, 
        response_type: Any, 
        qos: Optional[Dict[str, Any]] = None
    ) -> None:
        """
        Register a streaming service with security features.
        
        Args:
            service: Service name
            handler: Service handler
            request_type: Request message type
            response_type: Response message type
            qos: Quality of service parameters
        """
        # Not implementing security for streaming service registration yet
        # In a real implementation, we would unwrap incoming SecureMessage objects
        # and wrap outgoing responses
        await self.transport.register_stream_service(
            service, 
            handler, 
            request_type, 
            response_type, 
            qos
        )

File: ecps-go/pkg/trust/secure_transport.go
// Package trust provides security and trust mechanisms for the ECPS protocol stack.
package trust

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/ecps/ecps-go/pkg/core"
)

// SecureMessage wraps a message with security metadata
type SecureMessage struct {
	// Message is the original serialized message
	Message []byte `json:"message"`
	
	// MessageType is the type name of the original message
	MessageType string `json:"message_type"`
	
	// Signature is the message signature (if signed)
	Signature []byte `json:"signature,omitempty"`
	
	// EncryptedKey is the encrypted symmetric key (if encrypted)
	EncryptedKey []byte `json:"encrypted_key,omitempty"`
	
	// IV is the initialization vector for symmetric encryption
	IV []byte `json:"iv,omitempty"`
	
	// SecurityToken is the security token (e.g., JWT)
	SecurityToken string `json:"security_token,omitempty"`
	
	// SenderID is the ID of the sender
	SenderID string `json:"sender_id,omitempty"`
	
	// Timestamp is the message timestamp
	Timestamp int64 `json:"timestamp"`
}

// SecureTransport wraps a transport with security features
type SecureTransport struct {
	transport    core.Transport
	trustProvider *TrustProvider
	principalID   string
	securityToken string
}

// NewSecureTransport creates a new secure transport wrapper
func NewSecureTransport(transport core.Transport, trustProvider *TrustProvider) *SecureTransport {
	return &SecureTransport{
		transport:     transport,
		trustProvider: trustProvider,
	}
}

// SetPrincipal sets the principal ID and security token for the transport
func (t *SecureTransport) SetPrincipal(principalID string, securityToken string) {
	t.principalID = principalID
	t.securityToken = securityToken
}

// Start starts the transport
func (t *SecureTransport) Start(ctx context.Context) error {
	return t.transport.Start(ctx)
}

// Close closes the transport
func (t *SecureTransport) Close() error {
	return t.transport.Close()
}

// IsStarted checks if the transport is started
func (t *SecureTransport) IsStarted() bool {
	return t.transport.IsStarted()
}

// IsConnected checks if the transport is connected
func (t *SecureTransport) IsConnected() bool {
	return t.transport.IsConnected()
}

// Publish publishes a message with security features
func (t *SecureTransport) Publish(ctx context.Context, topic string, message interface{}, qos map[string]interface{}) error {
	// Skip security if trust level is none
	if t.trustProvider.trustLevel == TrustLevelNone {
		return t.transport.Publish(ctx, topic, message, qos)
	}
	
	// Get serializer from transport
	serializer, ok := t.transport.(interface{ Serializer() core.Serializer })
	if !ok {
		return errors.New("transport does not provide access to serializer")
	}
	
	// Serialize original message
	messageBytes, err := serializer.Serializer().Serialize(message, false)
	if err != nil {
		return fmt.Errorf("failed to serialize message: %w", err)
	}
	
	// Create secure message
	secureMsg := &SecureMessage{
		Message:     messageBytes,
		MessageType: fmt.Sprintf("%T", message),
		SenderID:    t.principalID,
		Timestamp:   time.Now().UnixNano(),
	}
	
	// Add security token if available
	if t.securityToken != "" {
		secureMsg.SecurityToken = t.securityToken
	}
	
	// Sign message if trust level requires it
	if t.trustProvider.trustLevel >= TrustLevelAuthenticated {
		signature, err := t.trustProvider.SignMessage(messageBytes)
		if err != nil {
			return fmt.Errorf("failed to sign message: %w", err)
		}
		secureMsg.Signature = signature
	}
	
	// Encrypt message if trust level requires it
	if t.trustProvider.trustLevel >= TrustLevelEncryption {
		// Generate a random symmetric key (AES-256)
		symmetricKey := make([]byte, 32)
		if _, err := rand.Read(symmetricKey); err != nil {
			return fmt.Errorf("failed to generate symmetric key: %w", err)
		}
		
		// Generate IV for AES-CBC
		iv := make([]byte, aes.BlockSize)
		if _, err := rand.Read(iv); err != nil {
			return fmt.Errorf("failed to generate IV: %w", err)
		}
		
		// Encrypt the message with AES-CBC
		block, err := aes.NewCipher(symmetricKey)
		if err != nil {
			return fmt.Errorf("failed to create AES cipher: %w", err)
		}
		
		// Pad the message to block size
		blockSize := aes.BlockSize
		paddingLength := blockSize - (len(messageBytes) % blockSize)
		paddedMessage := make([]byte, len(messageBytes)+paddingLength)
		copy(paddedMessage, messageBytes)
		for i := len(messageBytes); i < len(paddedMessage); i++ {
			paddedMessage[i] = byte(paddingLength)
		}
		
		// Encrypt with CBC mode
		mode := cipher.NewCBCEncrypter(block, iv)
		encryptedMessage := make([]byte, len(paddedMessage))
		mode.CryptBlocks(encryptedMessage, paddedMessage)
		
		// Encrypt the symmetric key with RSA public key (if available)
		if t.trustProvider.publicKey != nil {
			encryptedKey, err := rsa.EncryptOAEP(sha256.New(), rand.Reader, t.trustProvider.publicKey, symmetricKey, nil)
			if err != nil {
				return fmt.Errorf("failed to encrypt symmetric key: %w", err)
			}
			secureMsg.EncryptedKey = encryptedKey
		}
		
		secureMsg.Message = encryptedMessage
		secureMsg.IV = iv
	}
	
	// Publish secure message
	return t.transport.Publish(ctx, topic, secureMsg, qos)
}

// Subscribe subscribes to a topic with security features
func (t *SecureTransport) Subscribe(
	ctx context.Context,
	topic string,
	messageType interface{},
	handler interface{},
	qos map[string]interface{},
) error {
	// Skip security if trust level is none
	if t.trustProvider.trustLevel == TrustLevelNone {
		return t.transport.Subscribe(ctx, topic, messageType, handler, qos)
	}
	
	// Get serializer from transport
	serializer, ok := t.transport.(interface{ Serializer() core.Serializer })
	if !ok {
		return errors.New("transport does not provide access to serializer")
	}
	
	// Create secure message handler
	secureHandler := func(ctx context.Context, secureMsg *SecureMessage) error {
		// Verify signature if trust level requires it
		if t.trustProvider.trustLevel >= TrustLevelAuthenticated && secureMsg.Signature != nil {
			if err := t.trustProvider.VerifyMessage(secureMsg.Message, secureMsg.Signature); err != nil {
				return fmt.Errorf("message signature verification failed: %w", err)
			}
		}
		
		// Decrypt message if trust level requires it
		messageBytes := secureMsg.Message
		if t.trustProvider.trustLevel >= TrustLevelEncryption && secureMsg.EncryptedKey != nil {
			// Decrypt the symmetric key with our private key
			if t.trustProvider.privateKey == nil {
				return errors.New("no private key available for decryption")
			}
			
			symmetricKey, err := rsa.DecryptOAEP(sha256.New(), rand.Reader, t.trustProvider.privateKey, secureMsg.EncryptedKey, nil)
			if err != nil {
				return fmt.Errorf("failed to decrypt symmetric key: %w", err)
			}
			
			// Decrypt the message with AES-CBC
			if secureMsg.IV == nil || len(secureMsg.IV) != aes.BlockSize {
				return errors.New("invalid or missing IV for decryption")
			}
			
			block, err := aes.NewCipher(symmetricKey)
			if err != nil {
				return fmt.Errorf("failed to create AES cipher for decryption: %w", err)
			}
			
			if len(secureMsg.Message)%aes.BlockSize != 0 {
				return errors.New("encrypted message is not a multiple of block size")
			}
			
			mode := cipher.NewCBCDecrypter(block, secureMsg.IV)
			decryptedMessage := make([]byte, len(secureMsg.Message))
			mode.CryptBlocks(decryptedMessage, secureMsg.Message)
			
			// Remove padding
			if len(decryptedMessage) == 0 {
				return errors.New("decrypted message is empty")
			}
			
			paddingLength := int(decryptedMessage[len(decryptedMessage)-1])
			if paddingLength > len(decryptedMessage) || paddingLength == 0 {
				return errors.New("invalid padding in decrypted message")
			}
			
			messageBytes = decryptedMessage[:len(decryptedMessage)-paddingLength]
		}
		
		// Authenticate sender if trust level requires it
		var principal *Principal
		if t.trustProvider.trustLevel >= TrustLevelAuthenticated && secureMsg.SecurityToken != "" {
			var err error
			principal, err = t.trustProvider.ValidateJWT(secureMsg.SecurityToken)
			if err != nil {
				return fmt.Errorf("security token validation failed: %w", err)
			}
			
			// Check if sender ID matches the token subject
			if secureMsg.SenderID != "" && secureMsg.SenderID != principal.ID {
				return errors.New("sender ID does not match security token subject")
			}
		}
		
		// Authorize action if trust level requires it
		if t.trustProvider.trustLevel >= TrustLevelAuthorized && principal != nil {
			authorized, err := t.trustProvider.Authorize(ctx, principal, "subscribe", topic)
			if err != nil {
				return fmt.Errorf("authorization check failed: %w", err)
			}
			if !authorized {
				return errors.New("sender is not authorized to publish to this topic")
			}
		}
		
		// Deserialize original message
		origMsg, err := serializer.Serializer().Deserialize(messageBytes, messageType)
		if err != nil {
			return fmt.Errorf("failed to deserialize original message: %w", err)
		}
		
		// Add security context to the context
		if principal != nil {
			ctx = context.WithValue(ctx, "principal", principal)
		}
		
		// Call original handler with the deserialized message
		switch h := handler.(type) {
		case func(context.Context, interface{}) error:
			return h(ctx, origMsg)
		default:
			// Use reflection to call the handler with the correct type
			return errors.New("unsupported handler type")
		}
	}
	
	// Subscribe with secure handler
	return t.transport.Subscribe(ctx, topic, &SecureMessage{}, secureHandler, qos)
}

// Request sends a request with security features
func (t *SecureTransport) Request(
	ctx context.Context,
	service string,
	request interface{},
	timeout int64,
	qos map[string]interface{},
) (interface{}, error) {
	// Skip security if trust level is none
	if t.trustProvider.trustLevel == TrustLevelNone {
		return t.transport.Request(ctx, service, request, timeout, qos)
	}
	
	// Get serializer from transport
	serializer, ok := t.transport.(interface{ Serializer() core.Serializer })
	if !ok {
		return nil, errors.New("transport does not provide access to serializer")
	}
	
	// Serialize the request
	requestBytes, err := serializer.Serializer().Serialize(request)
	if err != nil {
		return nil, fmt.Errorf("failed to serialize request: %w", err)
	}
	
	// Create secure message
	secureMsg, err := t.createSecureMessage(requestBytes, request)
	if err != nil {
		return nil, fmt.Errorf("failed to create secure message: %w", err)
	}
	
	// Send secure request
	secureResponse, err := t.transport.Request(ctx, service, secureMsg, timeout, qos)
	if err != nil {
		return nil, err
	}
	
	// Unwrap secure response if it's a SecureMessage
	if secureMsg, ok := secureResponse.(*SecureMessage); ok {
		return t.unwrapSecureMessage(secureMsg, serializer.Serializer())
	}
	
	return secureResponse, nil
}

// StreamRequest sends a streaming request with security features
func (t *SecureTransport) StreamRequest(
	ctx context.Context,
	service string,
	request interface{},
	handler interface{},
	timeout int64,
	qos map[string]interface{},
) error {
	// Skip security if trust level is none
	if t.trustProvider.trustLevel == TrustLevelNone {
		return t.transport.StreamRequest(ctx, service, request, handler, timeout, qos)
	}
	
	// Get serializer from transport
	serializer, ok := t.transport.(interface{ Serializer() core.Serializer })
	if !ok {
		return errors.New("transport does not provide access to serializer")
	}
	
	// Create secure handler wrapper
	secureHandler := func(ctx context.Context, response interface{}) error {
		// Unwrap secure response if it's a SecureMessage
		if secureMsg, ok := response.(*SecureMessage); ok {
			unwrappedResponse, err := t.unwrapSecureMessage(secureMsg, serializer.Serializer())
			if err != nil {
				return fmt.Errorf("failed to unwrap secure response: %w", err)
			}
			
			// Call original handler with unwrapped response
			switch h := handler.(type) {
			case func(context.Context, interface{}) error:
				return h(ctx, unwrappedResponse)
			default:
				return errors.New("unsupported handler type")
			}
		} else {
			// Direct response, call handler as-is
			switch h := handler.(type) {
			case func(context.Context, interface{}) error:
				return h(ctx, response)
			default:
				return errors.New("unsupported handler type")
			}
		}
	}
	
	// Serialize the request
	requestBytes, err := serializer.Serializer().Serialize(request)
	if err != nil {
		return fmt.Errorf("failed to serialize request: %w", err)
	}
	
	// Create secure message
	secureMsg, err := t.createSecureMessage(requestBytes, request)
	if err != nil {
		return fmt.Errorf("failed to create secure message: %w", err)
	}
	
	// Send secure streaming request
	return t.transport.StreamRequest(ctx, service, secureMsg, secureHandler, timeout, qos)
}

// RegisterService registers a service with security features
func (t *SecureTransport) RegisterService(
	ctx context.Context,
	service string,
	handler interface{},
	requestType interface{},
	responseType interface{},
	qos map[string]interface{},
) error {
	// Skip security if trust level is none
	if t.trustProvider.trustLevel == TrustLevelNone {
		return t.transport.RegisterService(ctx, service, handler, requestType, responseType, qos)
	}
	
	// Get serializer from transport
	serializer, ok := t.transport.(interface{ Serializer() core.Serializer })
	if !ok {
		return errors.New("transport does not provide access to serializer")
	}
	
	// Create secure handler wrapper
	secureHandler := func(ctx context.Context, request interface{}) (interface{}, error) {
		// Unwrap secure request if it's a SecureMessage
		if secureMsg, ok := request.(*SecureMessage); ok {
			unwrappedRequest, err := t.unwrapSecureMessage(secureMsg, serializer.Serializer())
			if err != nil {
				return nil, fmt.Errorf("failed to unwrap secure request: %w", err)
			}
			
			// Call original handler with unwrapped request
			var response interface{}
			var handlerErr error
			
			switch h := handler.(type) {
			case func(context.Context, interface{}) (interface{}, error):
				response, handlerErr = h(ctx, unwrappedRequest)
			default:
				return nil, errors.New("unsupported handler type")
			}
			
			if handlerErr != nil {
				return nil, handlerErr
			}
			
			// Wrap response in secure message if not nil
			if response != nil {
				responseBytes, err := serializer.Serializer().Serialize(response)
				if err != nil {
					return nil, fmt.Errorf("failed to serialize response: %w", err)
				}
				
				secureResponse, err := t.createSecureMessage(responseBytes, response)
				if err != nil {
					return nil, fmt.Errorf("failed to create secure response: %w", err)
				}
				
				return secureResponse, nil
			}
			
			return nil, nil
		} else {
			// Direct request, call handler as-is
			switch h := handler.(type) {
			case func(context.Context, interface{}) (interface{}, error):
				return h(ctx, request)
			default:
				return nil, errors.New("unsupported handler type")
			}
		}
	}
	
	return t.transport.RegisterService(ctx, service, secureHandler, &SecureMessage{}, &SecureMessage{}, qos)
}

// RegisterStreamService registers a streaming service with security features
func (t *SecureTransport) RegisterStreamService(
	ctx context.Context,
	service string,
	handler interface{},
	requestType interface{},
	responseType interface{},
	qos map[string]interface{},
) error {
	// Skip security if trust level is none
	if t.trustProvider.trustLevel == TrustLevelNone {
		return t.transport.RegisterStreamService(ctx, service, handler, requestType, responseType, qos)
	}
	
	// Get serializer from transport
	serializer, ok := t.transport.(interface{ Serializer() core.Serializer })
	if !ok {
		return errors.New("transport does not provide access to serializer")
	}
	
	// Create secure streaming handler wrapper
	secureStreamingHandler := func(ctx context.Context, request interface{}, responseHandler func(context.Context, interface{}) error) error {
		// Create secure response handler
		secureResponseHandler := func(ctx context.Context, response interface{}) error {
			if response != nil {
				responseBytes, err := serializer.Serializer().Serialize(response)
				if err != nil {
					return fmt.Errorf("failed to serialize response: %w", err)
				}
				
				secureResponse, err := t.createSecureMessage(responseBytes, response)
				if err != nil {
					return fmt.Errorf("failed to create secure response: %w", err)
				}
				
				return responseHandler(ctx, secureResponse)
			}
			return responseHandler(ctx, nil)
		}
		
		// Unwrap secure request if it's a SecureMessage
		if secureMsg, ok := request.(*SecureMessage); ok {
			unwrappedRequest, err := t.unwrapSecureMessage(secureMsg, serializer.Serializer())
			if err != nil {
				return fmt.Errorf("failed to unwrap secure request: %w", err)
			}
			
			// Call original handler with unwrapped request
			switch h := handler.(type) {
			case func(context.Context, interface{}, func(context.Context, interface{}) error) error:
				return h(ctx, unwrappedRequest, secureResponseHandler)
			default:
				return errors.New("unsupported handler type")
			}
		} else {
			// Direct request, call handler as-is
			switch h := handler.(type) {
			case func(context.Context, interface{}, func(context.Context, interface{}) error) error:
				return h(ctx, request, responseHandler)
			default:
				return errors.New("unsupported handler type")
			}
		}
	}
	
	return t.transport.RegisterStreamService(ctx, service, secureStreamingHandler, &SecureMessage{}, &SecureMessage{}, qos)
}

// createSecureMessage creates a secure message with encryption and signing if required
func (t *SecureTransport) createSecureMessage(messageBytes []byte, originalMsg interface{}) (*SecureMessage, error) {
	secureMsg := &SecureMessage{
		Message:     messageBytes,
		MessageType: fmt.Sprintf("%T", originalMsg),
		SenderID:    t.trustProvider.principal.ID,
		Timestamp:   time.Now().UnixNano(),
	}
	
	// Add signature if trust level requires it
	if t.trustProvider.trustLevel >= TrustLevelAuthenticated {
		signature, err := t.trustProvider.SignMessage(messageBytes)
		if err != nil {
			return nil, fmt.Errorf("failed to sign message: %w", err)
		}
		secureMsg.Signature = signature
	}
	
	// Add encryption if trust level requires it
	if t.trustProvider.trustLevel >= TrustLevelEncryption {
		// Generate a random symmetric key (AES-256)
		symmetricKey := make([]byte, 32)
		if _, err := rand.Read(symmetricKey); err != nil {
			return nil, fmt.Errorf("failed to generate symmetric key: %w", err)
		}
		
		// Generate IV for AES-CBC
		iv := make([]byte, aes.BlockSize)
		if _, err := rand.Read(iv); err != nil {
			return nil, fmt.Errorf("failed to generate IV: %w", err)
		}
		
		// Encrypt the message with AES-CBC
		block, err := aes.NewCipher(symmetricKey)
		if err != nil {
			return nil, fmt.Errorf("failed to create AES cipher: %w", err)
		}
		
		// Pad the message to block size
		blockSize := aes.BlockSize
		paddingLength := blockSize - (len(messageBytes) % blockSize)
		paddedMessage := make([]byte, len(messageBytes)+paddingLength)
		copy(paddedMessage, messageBytes)
		for i := len(messageBytes); i < len(paddedMessage); i++ {
			paddedMessage[i] = byte(paddingLength)
		}
		
		// Encrypt with CBC mode
		mode := cipher.NewCBCEncrypter(block, iv)
		encryptedMessage := make([]byte, len(paddedMessage))
		mode.CryptBlocks(encryptedMessage, paddedMessage)
		
		// Encrypt the symmetric key with RSA public key (if available)
		if t.trustProvider.publicKey != nil {
			encryptedKey, err := rsa.EncryptOAEP(sha256.New(), rand.Reader, t.trustProvider.publicKey, symmetricKey, nil)
			if err != nil {
				return nil, fmt.Errorf("failed to encrypt symmetric key: %w", err)
			}
			secureMsg.EncryptedKey = encryptedKey
		}
		
		secureMsg.Message = encryptedMessage
		secureMsg.IV = iv
	}
	
	return secureMsg, nil
}

// unwrapSecureMessage unwraps a secure message, verifying signature and decrypting if needed
func (t *SecureTransport) unwrapSecureMessage(secureMsg *SecureMessage, serializer core.Serializer) (interface{}, error) {
	messageBytes := secureMsg.Message
	
	// Verify signature if present
	if secureMsg.Signature != nil {
		if err := t.trustProvider.VerifyMessage(messageBytes, secureMsg.Signature); err != nil {
			return nil, fmt.Errorf("signature verification failed: %w", err)
		}
	}
	
	// Decrypt if encrypted
	if secureMsg.EncryptedKey != nil && secureMsg.IV != nil {
		// Decrypt the symmetric key with our private key
		if t.trustProvider.privateKey == nil {
			return nil, errors.New("no private key available for decryption")
		}
		
		symmetricKey, err := rsa.DecryptOAEP(sha256.New(), rand.Reader, t.trustProvider.privateKey, secureMsg.EncryptedKey, nil)
		if err != nil {
			return nil, fmt.Errorf("failed to decrypt symmetric key: %w", err)
		}
		
		// Decrypt the message with AES-CBC
		if len(secureMsg.IV) != aes.BlockSize {
			return nil, errors.New("invalid IV size for decryption")
		}
		
		block, err := aes.NewCipher(symmetricKey)
		if err != nil {
			return nil, fmt.Errorf("failed to create AES cipher for decryption: %w", err)
		}
		
		if len(secureMsg.Message)%aes.BlockSize != 0 {
			return nil, errors.New("encrypted message is not a multiple of block size")
		}
		
		mode := cipher.NewCBCDecrypter(block, secureMsg.IV)
		decryptedMessage := make([]byte, len(secureMsg.Message))
		mode.CryptBlocks(decryptedMessage, secureMsg.Message)
		
		// Remove padding
		if len(decryptedMessage) == 0 {
			return nil, errors.New("decrypted message is empty")
		}
		
		paddingLength := int(decryptedMessage[len(decryptedMessage)-1])
		if paddingLength > len(decryptedMessage) || paddingLength == 0 {
			return nil, errors.New("invalid padding in decrypted message")
		}
		
		messageBytes = decryptedMessage[:len(decryptedMessage)-paddingLength]
	}
	
	// Deserialize the original message
	return serializer.Deserialize(messageBytes, nil)
}